Config switch to turn on/off Quic processing

nezdolik opened the pull request:
Signed-off-by: Kateryna Nezdolii <nezdolik@spotify.com>

Description:
Extending quic listener configuration with runtime feature flag for enabling/disabling quic processing. Flag controls onReadReady() and onWriteReady() callback methods and is enabled by default.
Risk Level: Low/Medium
Testing: Unit tests+Integration tests.
Docs Changes: NA?
Release Notes:
Fixes ##9230

nezdolik commented on api/envoy/api/v2/listener/udp_listener_config.proto:
```diff
@@ -30,6 +30,17 @@ message UdpListenerConfig {
     google.protobuf.Any typed_config = 3;
   }
+  
+  enum UdpShutDownMode {
```
cc @alyssawilk @danzh2010 @mattklein123 this is initial draft for config layout.

htuch commented on api/envoy/api/v2/listener/udp_listener_config.proto:
```diff
@@ -30,6 +30,17 @@ message UdpListenerConfig {
     google.protobuf.Any typed_config = 3;
   }
+  
+  enum UdpShutDownMode {
```
You should probably have some `0` entry such as `UNSET`.

nezdolik replied:
@htuch was thinking about it, but it could lead to interesting state. What would be the logic when `udp_enabled` is false (we want to stop udp processing) and shutdown_mode is unset?

mattklein123 replied:
I would probably just make the 0 entry the default which is graceful?

htuch replied:
Yeah, that is probably a reasonable default that won't change, let's do that.

mattklein123 commented on api/envoy/api/v2/listener/udp_listener_config.proto:
```diff
@@ -30,6 +30,17 @@ message UdpListenerConfig {
+
+  UdpShutDownMode shutdown_mode = 4 [(validate.rules).enum = {defined_only: true}];
+
+  bool udp_enabled = 5;
```
Sorry I'm a little confused about what is being proposed here. Do you mind adding some more comments?

htuch replied:
+!

nezdolik replied:
@mattklein123 @htuch for turning off/on udp/quick processing, my proposal is to have boolean switch on udp listener config + shutdown mode field (for shutdown type). Submitted this draft config to get community agreement on the approach, code will follow.

htuch replied:
Why don't we just remove the listener and drain as normal? Trying to grok what is distinct here vs. any other listener that comes to life at some point, starts accepting, and then is removed and drained.

nezdolik replied:
@htuch thanks for suggestion, i did not work much with LDS, so missed that opportunity. Agree that this is cleaner way, however currently listener manager supports only draining (graceful shutdown) of active listeners. In case we go with LDS approach, it will require code change to support multiple draining modes, right?

htuch replied:
Yes, but I think a drain timeout could subsume this; if you provide a drain timeout and make it zero, that will give you an immediate drain, > 0 will give you a graceful drain.

nezdolik replied:
@htuch i see. But if to read [issue](https://github.com/envoyproxy/envoy/issues/9230) description, it says `Current QUIC support can only be turned on/off via LDS. We can do better here by adding an Envoy reloadable feature for QUIC to turn the switch via runtime dynamic config push.` As i understood it meant introducing config switch.

alyssawilk replied:
@htuch if we have UDP proxy and QUIC share a listener (443, select which via VIP) and want to turn off one and not the other, how do you do that without explicit config?  If there's existing framework to do that than I'm good.

htuch replied:
I guess this comes down to documentation. What is this `udp_enabled` supposed to do?  Does it do both UDP and QUIC? What does the drain do? Is it just for UDP or also for QUIC? Is there a parallel way to achieve the same ends for QUIC? 

Maybe @danzh2010 can weight in on what the goal is.

mattklein123 replied:
Would it be better to just have a per-UDP listener that has a runtime kill switch that will just stop accepting packets? I think maybe this is a simpler way of handling the intention?

alyssawilk replied:
I guess it depends how we structure shared listeners.
I thought if we had QUIC and QUIC-proxy listening on UDP/443 they'd share a listener, and some higher level filter would determine which packets went where.  In that case as said it'd be nice to able to disable them one at a time.  We can deal without but it's a nice to have.
If we think we'd have a QUIC/443 listener and a QUIC-proxy/443 listener just sharing incoming traffic via some combination of SO_REUSEPORT and BPF this would be fine.
I'd assumed we were going to do more of the former, at which point a non-listener kill switch has value IMO (though again we can live without it)

mattklein123 replied:
My actual assumption (potentially naive) was that QUIC and QUIC-proxy would probably be different deployments of the binary in different auto scaling groups, etc. What is the use case in which you envision the same process doing both?

alyssawilk replied:
Generic example is for edge proxies with low-value certs that do transparent proxying for latency gains.  QUIC-terminate locally served content, UDP proxy where you'd TCP proxy because DNS SRV doesn't work.
We have two different frontline deployments which do TLS / TCP proxy based on VIP and one of them did QUIC+Udp-proxy for years.
We haven't had an issue where we needed to turn off one and not the other yet, but as tunneling over QUIC picks up steam (and that team has agreed to OSS once we land core QUIC) I think it gets more useful to turn different modes down separately.
We can always start simple and add more later though!

mattklein123 replied:
> We can always start simple and add more later though!

That would be my feeling. Right now we don't have any filter chain concept for UDP (I would like to fix that as part of a general listener API cleanup), and without that I'm not sure how we would do something like ^ today. IMO right now I think we could just have a per-listener runtime kill switch for UDP listeners? Or even just for the QUIC listener config itself?

nezdolik replied:
Thanks for feedback, it's valuable information to get familiar with. So we agreed to introduce separate config switches for upd and quic listeners.

nezdolik commented:
As for now i see that ActiveRawUdpListenerConfig and ActiveQuicListenerConfig are not related in terms of type inheritance and share no similar props apart from name, so question is do we want to support shutdown flag for both listeners or only quic. If for both it will require some code refactoring.

nezdolik commented on source/server/active_raw_udp_listener_config.cc:
```diff
@@ -24,7 +24,9 @@ ProtobufTypes::MessagePtr ActiveRawUdpListenerConfigFactory::createEmptyConfigPr
 ActiveRawUdpListenerConfigFactory::createActiveUdpListenerFactory(
-    const Protobuf::Message& /*message*/) {
+    const Protobuf::Message& message) {
+  auto& config =
```
for now this cast will not work, need to reorg udp listener proto config to have one top level message

mattklein123 commented on api/envoy/api/v2/listener/quic_config.proto:
```diff
@@ -31,3 +32,20 @@ message QuicProtocolOptions {
   google.protobuf.Duration crypto_handshake_timeout = 3;
 }
+
+message ActiveQuicListenerConfig {
```
I would probably just name this `QuicListenerConfig`. I think we can drop the Active (having it in the other place is a mistake IMO).

In terms of the shutoff, what I'm proposing is just having a `RuntimeFeatureFlag` which says whether to process packets or not. I think this is ultimately the kill switch we are looking for? @alyssawilk @danzh2010  WDYT?

alyssawilk replied:
Again I think it's not the switch I'm looking for in the long term (because not granular enough) but I'm fine landing whatever controls we think are useful today and we can add the finicky ones once we're running in prod and care more :-)

mattklein123 commented on api/envoy/api/v2/listener/udp_listener_config.proto:
```diff
@@ -33,4 +34,17 @@ message UdpListenerConfig {
 }
 
 message ActiveRawUdpListenerConfig {
+  enum UdpShutDownMode {
```
I'm not completely positive we want to bother with this kill switch for raw UDP right now. Maybe defer that for later? If we do it here, we should de-dup the messages and have a common configuration message of some type.

nezdolik commented:
Spent most of the time on task, which may be out of scope for this change, basically trying to propagate Envoy primitives (like Runtime) all the way down from `server instance->worker->connection handler->active quic listener factory->active quic listener1, which introduced cyclic dependency in bazel modules. As current code is not production ready (and ActiveQuicListenerFactory not being instantiated from conn_handler), change may be fine as it is (adding `Runtime` field to ActiveQuicListenerFactory), but in the long run code needs to be refactored, possibly introducing UdpFactoryContext that would provide to resources like Dispatcher, Runtime::Loader etc.
Still need to complete tests.
Would appreciate opinions if refactoring change should be part of this task or not.

mattklein123 commented:
Check out `Runtime::LoaderSingleton::getExisting()` which will probably make this task easier. This is one case where we break our rule about no statics.

nezdolik commented:
took me some to understand that `bazel build //source/exe:envoy-static` is not including `//source/extensions/quic_listeners/quiche:...` :)

nezdolik commented on source/extensions/quic_listeners/quiche/BUILD:
```diff
@@ -275,6 +275,7 @@ envoy_cc_library(
         "//source/common/protobuf:utility_lib",
         "//source/server:connection_handler_lib",
         "@envoy_api//envoy/config/listener/v3:pkg_cc_proto",
+        "//source/common/runtime:runtime_lib",
```
may need to sort alphabetically

nezdolik commented on source/extensions/quic_listeners/quiche/active_quic_listener.h:
```diff
@@ -83,16 +90,17 @@ class ActiveQuicListenerFactory : public Network::ActiveUdpListenerFactory {
   friend class ActiveQuicListenerFactoryPeer;
 
   quic::QuicConfig quic_config_;
+  envoy::config::core::v3::RuntimeFeatureFlag enabled;
```
oh, should have suffixed with underscore

nezdolik commented:
@mattklein123 @danzh2010 please take a look. There is failing test in quiche `QuicMemSliceStorageImplTest`, did not find location (public github repo) where to report issues for quiche.

mattklein123 commented on source/extensions/quic_listeners/quiche/active_quic_listener.cc:
```diff
@@ -94,16 +100,24 @@ void ActiveQuicListener::onData(Network::UdpRecvData& data) {
+  if (!enabled()) {
+    ENVOY_LOG(trace, "Quic listener {}: runtime disabled", config_.name());
+    return;
+  }
```
I think we can probably just guard the read ready case? I will defer to @danzh2010 on this.

danzh2010 replied:
Yes, guarding onData()not necessary. 

On the other hand, do we want the kill switch to gracefully close all new and existing connections before blocking all the incoming packets? If no, I think guarding at onReadReady() and onWriteReady() is sufficient. Otherwise we probably need to use QUICHE interface in QuicDispatcher to reset all the quic connections. @mattklein123 @alyssawilk

nezdolik commented on api/envoy/api/v2/listener/quic_config.proto:
```diff
@@ -30,4 +32,8 @@ message QuicProtocolOptions {
   // 20000ms if not specified.
   google.protobuf.Duration crypto_handshake_timeout = 3;
+
+  // Runtime flag that controls whether the listener is enabled or not. If not specified, defaults
```
may need to add test when property is not specified

nezdolik commented:
@danzh2010 updated description for PR

mattklein123 commented:
> On the other hand, do we want the kill switch to gracefully close all new and existing connections before blocking all the incoming packets? If no, I think guarding at onReadReady() and onWriteReady() is sufficient. Otherwise we probably need to use QUICHE interface in QuicDispatcher to reset all the quic connections. @mattklein123 @alyssawilk

IMO we don't need to be graceful here so I would go with the simple solution of just blocking read packets.

htuch commented on api/envoy/api/v2/listener/quic_config.proto:
```diff
@@ -30,4 +32,8 @@ message QuicProtocolOptions {
+
+  // Runtime flag that controls whether the listener is enabled or not. If not specified, defaults
+  // to enabled.
+  core.RuntimeFeatureFlag enabled = 4;
```
Just chiming in here late, but curious why we needed the ability to runtime enable UDP on a per-listener basis, i.e. why it's not sufficient to do runtime override proxy-wide. I'm guessing most of the time you want this that something has gone seriously wrong serving UDP and you want everyone to fall back to TCP, and that would be proxy-wide. I can see the per-listener argument in a multi-tenant scenario, but I imagine in that case you would be willing to accept an LDS push.

Otherwise API LGTM.

nezdolik replied:
@htuch this was agreed upon as optimal approach based on effort level and current state of UDP listeners codebase.  Please also see long discussion here: https://github.com/envoyproxy/envoy/pull/9679#discussion_r369872379

htuch replied:
OK, resolving, sorry for the redundant thread.

nezdolik commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -221,16 +244,22 @@ TEST_P(ActiveQuicListenerTest, FailSocketOptionUponCreation) {
+      ActiveQuicListenerPeer::quic_dispatcher(*quic_listener_);
+  quic::QuicBufferedPacketStore* const buffered_packets =
+      quic::test::QuicDispatcherPeer::GetBufferedPackets(envoy_quic_dispatcher);
+  ConfigureMocks(/* connection_count = */ 1, /* runtime_enabled = */ true);
```
@danzh2010 it was intentional to unify expectations across all tests and to make it easier to understand what is outcome of given test input (like for example, if quic is enabled and we send some data to listener, there should be sessions created), even though is can be obvious. wdyt?

danzh2010 replied:
Okay. I'm fine with it. The intention wasn't obvious to me at the first glance.

nezdolik commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -256,6 +285,7 @@ TEST_P(ActiveQuicListenerTest, ProcessBufferedChlos) {
   EXPECT_TRUE(buffered_packets->HasBufferedPackets(
       quic::test::TestConnectionId(ActiveQuicListener::kNumSessionsToCreatePerLoop + 1)));
   EXPECT_TRUE(buffered_packets->HasChlosBuffered());
+  EXPECT_FALSE(envoy_quic_dispatcher->session_map().empty());
```
same as above

nezdolik commented:
> > On the other hand, do we want the kill switch to gracefully close all new and existing connections before blocking all the incoming packets? If no, I think guarding at onReadReady() and onWriteReady() is sufficient. Otherwise we probably need to use QUICHE interface in QuicDispatcher to reset all the quic connections. @mattklein123 @alyssawilk
> 
> IMO we don't need to be graceful here so I would go with the simple solution of just blocking read packets.

I may be lacking some knowledge here, but why not to also guard onData()? OnData() implementation in quic dispatcher does bunch of things like buffering of packages if applicable, would not we want not to do it if quic processing is disabled @danzh2010 @mattklein123 ?

mattklein123 commented:
> I may be lacking some knowledge here, but why not to also guard onData()? OnData() (aka ProcessPacket(()) implementation in quic dispatcher does bunch of things like buffering of packages if applicable, would not we want not to do it if quic processing is disabled @danzh2010 @mattklein123 ?

I will defer to @danzh2010 as the expert, but we only need to hard block the receipt of packets. Everything else will eventually time out.

nezdolik commented:
@danzh2010 @mattklein123 applied review comments. Just one thing, when feature flag is not specified in config, the default value is false, which contradicts to this comment:
https://github.com/envoyproxy/envoy/pull/9679/files#diff-4c87225965fc01abe93665952e63ada9R36

nezdolik commented:
Please ignore my previous comment, the test was checking default value, but not the actual value of the flag. Did a quick check and runtime feature is enabled by default if not specified in config.

danzh2010 commented:
> > > On the other hand, do we want the kill switch to gracefully close all new and existing connections before blocking all the incoming packets? If no, I think guarding at onReadReady() and onWriteReady() is sufficient. Otherwise we probably need to use QUICHE interface in QuicDispatcher to reset all the quic connections. @mattklein123 @alyssawilk
> > 
> > 
> > IMO we don't need to be graceful here so I would go with the simple solution of just blocking read packets.
> 
> I may be lacking some knowledge here, but why not to also guard onData()? OnData() (aka ProcessPacket(()) implementation in quic dispatcher does bunch of things like buffering of packages if applicable, would not we want not to do it if quic processing is disabled @danzh2010 @mattklein123 ?

onData() is the callback  in onReadReady(). If you already guarded there, onData() won't be reached at all. You can add a temporary ASSERT in onData() to check it.

nezdolik commented:
> > > > On the other hand, do we want the kill switch to gracefully close all new and existing connections before blocking all the incoming packets? If no, I think guarding at onReadReady() and onWriteReady() is sufficient. Otherwise we probably need to use QUICHE interface in QuicDispatcher to reset all the quic connections. @mattklein123 @alyssawilk
> > > 
> > > 
> > > IMO we don't need to be graceful here so I would go with the simple solution of just blocking read packets.
> > 
> > 
> > I may be lacking some knowledge here, but why not to also guard onData()? OnData() (aka ProcessPacket(()) implementation in quic dispatcher does bunch of things like buffering of packages if applicable, would not we want not to do it if quic processing is disabled @danzh2010 @mattklein123 ?
> 
> onData() is the callback in onReadReady(). If you already guarded there, onData() won't be reached at all. You can add a temporary ASSERT in onData() to check it.

Thanks for explanation @danzh2010.

nezdolik commented:
/rebuild

danzh2010 commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -270,5 +300,16 @@ TEST_P(ActiveQuicListenerTest, ProcessBufferedChlos) {
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  // If listener was enabled, there should have been session created for active connection.
+  EXPECT_TRUE(envoy_quic_dispatcher->session_map().empty());
+  ReadFromClientSockets();
```
This is not necessary as the listener is black-holing all the packets.

nezdolik replied:
fixed

danzh2010 commented:
> Please ignore my previous comment, the test was checking default value, but not the actual value of the flag. Did a quick check and runtime feature is enabled by default if not specified in config.

Thanks for checking the default behavior! Do you mind add the above QuicEnabledByDefault test somewhere to enforce this?

nezdolik commented:
@danzh2010 believe all review comments are addressed. I had to fix additional clang_tidy errors, that were not popping up in CI before. (apparently something changed in CI)

nezdolik commented:
@mattklein123 could you kindly retrigger the build, there is some redis test failing. It should not be related to the proposed change

mattklein123 commented on source/extensions/quic_listeners/quiche/active_quic_listener.cc:
```diff
@@ -93,10 +95,18 @@ void ActiveQuicListener::onData(Network::UdpRecvData& data) {
 }
 
 void ActiveQuicListener::onWriteReady(const Network::Socket& /*socket*/) {
+  if (!enabled()) {
+    ENVOY_LOG(trace, "Quic listener {}: runtime disabled", config_.name());
+    return;
+  }
```
I think for simplicity we can remove this check? It doesn't matter I think if we get a few last writes out, as long as we don't do any more reads? @danzh2010 wdyt?

danzh2010 replied:
I would think so too, but either way is fine for me.

nezdolik replied:
ok, let me remove the check

htuch commented:
Please merge master to pick up https://github.com/envoyproxy/envoy/pull/10672. We no longer accept changes to v2 (without explicit exception), so any API modifications should happen in v3. If this PR is adding a new proto, please follow the updated instructions in https://github.com/envoyproxy/envoy/blob/master/api/STYLE.md#adding-an-extension-configuration-to-the-api.

danzh2010 commented on test/extensions/quic_listeners/quiche/active_quic_listener_config_test.cc:
```diff
@@ -29,6 +33,9 @@ TEST(ActiveQuicListenerConfigTest, CreateActiveQuicListenerFactory) {
     idle_timeout: {
       seconds: 2
     }
+    enabled:
+      default_value: true
+      runtime_key: foo_key
```
What would happen if you leave this out? What's the default key and value?

nezdolik replied:
@danzh2010 on this test level is not possible to verify value of feature flag, as we are only testing how yaml is parsed in corresponding code/proto representaion, and proto for RuntimeFeatureFlag looks like this:
```
message RuntimeFeatureFlag {
  google.protobuf.BoolValue default_value = 1 [(validate.rules).message = {required: true}];
  string runtime_key = 2 [(validate.rules).string = {min_bytes: 1}];
}
```
Actual value of feature flag is obtained from `Runtime` primitive (which is available in active_quic_listener test). Absence of feature flag configuration is covered by test: 
https://github.com/envoyproxy/envoy/pull/9679/files#diff-6a27a42427472c010c1a55f272fa72f4R272

nezdolik commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -221,30 +246,52 @@ TEST_P(ActiveQuicListenerTest, FailSocketOptionUponCreation) {
 TEST_P(ActiveQuicListenerTest, ReceiveFullQuicCHLO) {
-  ConfigureMocks(/* connection_count = */ 1);
-  SendFullCHLO(quic::test::TestConnectionId(1));
+  EnvoyQuicDispatcher* const envoy_quic_dispatcher =
```
good catch

mattklein123 commented:
Sorry needs a master merge.

/wait

nezdolik commented:
@mattklein123 added a test that i expect to fail: https://github.com/envoyproxy/envoy/pull/9679/files#diff-1d14c854240d676b61a3cb2b958df06fR79
Wanted to confirm that when feature flag is not specified, it will have empty runtime key and "true" default value. (This was to address @danzh2010 question).
There will be a one line test code correction needed after this.
Apologies for spamming CI pipeline, i could have tested it locally in Docker, but is very slow.

danzh2010 commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -221,30 +249,46 @@ TEST_P(ActiveQuicListenerTest, FailSocketOptionUponCreation) {
-  EnvoyQuicDispatcher* const envoy_quic_dispatcher =
-      ActiveQuicListenerPeer::quic_dispatcher(*quic_listener_);
+/* Not configuring runtime key for `quic.enabled`
+and checking that quic processing is enabled by default. */
```
I think we are still configuring quic.enabled as long as enabledFlag() is still called in SetUp(). This test is testing the default behavior of Runtime::MockLoader.

danzh2010 commented:
> @mattklein123 added a test that i expect to fail: https://github.com/envoyproxy/envoy/pull/9679/files#diff-1d14c854240d676b61a3cb2b958df06fR79
> Wanted to confirm that when feature flag is not specified, it will have empty runtime key and "true" default value. (This was to address @danzh2010 question).
> There will be a one line test code correction needed after this.
> Apologies for spamming CI pipeline, i could have tested it locally in Docker, but is very slow.

Thanks for adding the test! Doesn't the new test fail locally when you ran blaze test?

nezdolik commented:
> > @mattklein123 added a test that i expect to fail: https://github.com/envoyproxy/envoy/pull/9679/files#diff-1d14c854240d676b61a3cb2b958df06fR79
> > Wanted to confirm that when feature flag is not specified, it will have empty runtime key and "true" default value. (This was to address @danzh2010 question).
> > There will be a one line test code correction needed after this.
> > Apologies for spamming CI pipeline, i could have tested it locally in Docker, but is very slow.
> 
> Thanks for adding the test! Doesn't the new test fail locally when you ran blaze test?

@danzh2010 I was never able to run quic tests locally on MacOs, think i pasted initial error long time ago in quic channel. Nowadays it fails with long verbose error that's not obvious to me. Will send most recent full log in quic channel.

nezdolik commented:
@danzh2010 @mattklein123 removed the test that was relying on default mock behaviour, my bad i missed that one. 
There was a question like "will be quic processing working if  there is no configuration for it in proto". Currently we cannot really do end-to-end test because of using mocks, but looking into code, there are multiple possible scenarios:

1. There is no configuration of that feature flag, in that case runtime_key string will be empty and code will return false (there is code part that checks if key is empty and if it is it returns false).
https://github.com/envoyproxy/envoy/blob/master/source/common/runtime/runtime_impl.cc#L269

2. There is config for feature flag, runtime_key is specified and in that case it would be responsibility of user which default value do they set (default_value is required field in that proto).

It could either be responsibility of user to provide configuration for that flag or we could add some code guard like “if runtime_key is empty set default_value to true”.

danzh2010 commented:
> 1. There is no configuration of that feature flag, in that case runtime_key string will be empty and code will return false (there is code part that checks if key is empty and if it is it returns false).
>    https://github.com/envoyproxy/envoy/blob/master/source/common/runtime/runtime_impl.cc#L269

I think you looked at the wrong caller of getBoolean(), Runtime::FeatureFlag::default_value_ is true if not configured:
https://github.com/envoyproxy/envoy/blob/15330aa60bd6da79be02f00b8130b527f11eb0fa/source/common/runtime/runtime_protos.h#L20

nezdolik commented on test/extensions/quic_listeners/quiche/active_quic_listener_config_test.cc:
```diff
@@ -41,6 +48,35 @@ TEST(ActiveQuicListenerConfigTest, CreateActiveQuicListenerFactory) {
+  EXPECT_EQ("foo_key", runtime_enabled.runtime_key());
+}
+
+TEST(ActiveQuicListenerConfigTest, QuicListenerFlagNotConfigured) {
```
should we possibly remove this test, is there much value? (testing protobuf default values)

nezdolik commented:
@danzh2010 in that case we are good to merge PR? Do you have any other review comments? There is one test (added a comment about it), that does not bring much value, could be removed.

nezdolik commented:
/rebuild

nezdolik commented:
/retest

danzh2010 commented on test/extensions/quic_listeners/quiche/active_quic_listener_config_test.cc:
```diff
@@ -41,6 +48,35 @@ TEST(ActiveQuicListenerConfigTest, CreateActiveQuicListenerFactory) {
+  envoy::config::core::v3::RuntimeFeatureFlag& runtime_enabled =
+      ActiveQuicListenerFactoryPeer::runtimeEnabled(
+          dynamic_cast<ActiveQuicListenerFactory&>(*listener_factory));
+  EXPECT_EQ(false, runtime_enabled.default_value().value());
```
Can you check runtime_enabled.has_default_value() here instead? If it's false, we know that FeatureFlags will be instantiated with default value true.

nezdolik replied:
fixed

danzh2010 commented on source/extensions/quic_listeners/quiche/active_quic_listener.cc:
```diff
@@ -112,7 +118,7 @@ void ActiveQuicListener::shutdownListener() {
 ActiveQuicListenerFactory::ActiveQuicListenerFactory(
     const envoy::config::listener::v3::QuicProtocolOptions& config, uint32_t concurrency)
-    : concurrency_(concurrency) {
+    : concurrency_(concurrency), enabled_(config.enabled()) {
```
what if config.has_enabled() == false?

nezdolik replied:
Just verified the implementation, has_enabled() checks if field is not (either nullptr or is set to default value), but enabled instead always provides value to the caller:
```
inline const ::envoy::config::core::v3::RuntimeFeatureFlag& QuicProtocolOptions::_internal_enabled() const {
  const ::envoy::config::core::v3::RuntimeFeatureFlag* p = enabled_;
  return p != nullptr ? *p : *reinterpret_cast<const ::envoy::config::core::v3::RuntimeFeatureFlag*>(
      &::envoy::config::core::v3::_RuntimeFeatureFlag_default_instance_);
}
```
If config is empty, enabled field is set to default instance.

danzh2010 commented on test/extensions/quic_listeners/quiche/active_quic_listener_config_test.cc:
```diff
@@ -41,6 +48,35 @@ TEST(ActiveQuicListenerConfigTest, CreateActiveQuicListenerFactory) {
+      seconds: 2
+    }
+  )EOF";
+  TestUtility::loadFromYaml(yaml, *config);
```
I'd like to verify that config.has_enabled() returns false here. Otherwise TestUtility::loadFromYaml() might be providing a default FeatureFlag message here.

nezdolik replied:
added now

nezdolik commented:
@danzh2010 switched to real loader in test, will address review comments soon.

nezdolik commented:
@danzh2010 switched to real Runtime object and added end to end test for empty config. Please take a look, i believe it addresses your review comments.

danzh2010 commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -60,13 +76,21 @@ class ActiveQuicListenerTest : public testing::TestWithParam<Network::Address::I
 
-    quic_listener_ = std::make_unique<ActiveQuicListener>(
-        *dispatcher_, connection_handler_, listen_socket_, listener_config_, quic_config_, nullptr);
+    quic_listener_ = std::make_unique<ActiveQuicListener>(*dispatcher_, connection_handler_,
+                                                          listen_socket_, listener_config_,
+                                                          quic_config_, nullptr, enabledFlag());
```
Overriding enabledFlag() is a bit indirect. Can you add two methods like below:
```   
std::make_unique<ActiveQuicListener> createQuicListenerFactory(const std::string& yaml) {
std::string listener_name = QuicListenerName;
    auto& config_factory =
        Config::Utility::getAndCheckFactoryByName<Server::ActiveUdpListenerConfigFactory>(
            listener_name);
    ProtobufTypes::MessagePtr config = config_factory.createEmptyConfigProto();
    TestUtility::loadFromYaml(yaml, *config);
    return config_factory.createActiveUdpListenerFactory(*config, /*concurrency=*/1);

virtual std::string yamlForQuicConfig() {
  return  R"EOF(
runtime_key: "quic.enabled"
default_value: true
)EOF";
}
```
And use them here:
```
EXPECT_CALL(listener_config_, listenSocketFactory());
EXPECT_CALL(listener_config_. socket_factory_, getListenSocket).WillOnce(Return(listen_socket_));
quic_listener_ = createQuicListenerFactory(yamlForQuicConfig())->createActiveUdpListener(*dispatcher_, connection_handler_, ...);
```

In ActiveQuicListenerEmptyFlagConfigTest, 
```
std::string yamlForQuicConfig() override {
 return R"EOF(
    max_concurrent_streams: 10
  )EOF";
}
```

danzh2010 commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -270,5 +319,65 @@ TEST_P(ActiveQuicListenerTest, ProcessBufferedChlos) {
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// Quic listener should be enabled by default, if not enabled explicitly in config.
```
Thanks for adding this test!

nezdolik commented on bazel/envoy_internal.bzl:
```diff
@@ -7,7 +7,7 @@ def envoy_copts(repository, test = False):
         "-Wall",
         "-Wextra",
-        "-Werror",
+        #"-Werror",
```
@danzh2010 i was getting clang complaint for quiche library code, something like "comparing signed and unsigned ...". Is just for running builds locally, should not have gotten into pr

danzh2010 replied:
I'm fixing this issue in quiche now. It will be updated soon.

nezdolik commented on source/extensions/quic_listeners/quiche/active_quic_listener.h:
```diff
@@ -50,6 +54,8 @@ class ActiveQuicListener : public Network::UdpListenerCallbacks,
   void resumeListening() override;
   void shutdownListener() override;
 
+  bool enabled() { return enabled_.enabled(); }
```
@danzh2010 alright, just looked up that method in other envoy modules and followed by example.

danzh2010 replied:
This is not the only definition of quic listener being enabled. I would prefer not to add this interface.

nezdolik commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -59,17 +78,48 @@ class ActiveQuicListenerTest : public testing::TestWithParam<Network::Address::I
 
-    quic_listener_ = std::make_unique<ActiveQuicListener>(
-        *dispatcher_, connection_handler_, listen_socket_, listener_config_, quic_config_, nullptr);
+    EXPECT_CALL(listener_config_, listenSocketFactory())
```
this is needed to wire up mocks, that are used inside listener factory when creating quic listener

mattklein123 replied:
For this type of thing it's better to use ON_CALL(...).WillByDefault()

nezdolik commented:
There is failing coverage: https://343614-65214191-gh.circle-artifacts.com/0/coverage/index.html but none if it is related to this PR. Going to merge master and see if it helps

mattklein123 commented on source/extensions/quic_listeners/quiche/active_quic_listener.h:
```diff
@@ -25,12 +27,14 @@ class ActiveQuicListener : public Network::UdpListenerCallbacks,
                      Network::ListenerConfig& listener_config, const quic::QuicConfig& quic_config,
-                     Network::Socket::OptionsSharedPtr options);
+                     Network::Socket::OptionsSharedPtr options,
+                     const envoy::config::core::v3::RuntimeFeatureFlag enabled);
```
nit: pass by const ref here and below

nezdolik replied:
fixed

danzh2010 commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -270,5 +348,55 @@ TEST_P(ActiveQuicListenerTest, ProcessBufferedChlos) {
+  EXPECT_FALSE(ActiveQuicListenerPeer::enabled(*quic_listener_));
+}
+
+TEST_P(ActiveQuicListenerTest, QuicProcessingDisabledAndEnabled) {
```
Add `EXPECT_TRUE(ActiveQuicListenerPeer::enabled(*quic_listener_));` before flipping `quic.enabled`

nezdolik replied:
done

nezdolik commented on test/extensions/quic_listeners/quiche/active_quic_listener_test.cc:
```diff
@@ -270,5 +348,55 @@ TEST_P(ActiveQuicListenerTest, ProcessBufferedChlos) {
   ReadFromClientSockets();
 }
 
+TEST_P(ActiveQuicListenerTest, QuicProcessingDisabled) {
```
removed

nezdolik commented:
```FAILED_PRECONDITION: there are no bots capable of executing....``` in ci, need to wait for fix...

nezdolik commented:
@danzh2010 @mattklein123 applied review comments.
